fairscale/optim/oss.py

```diff
@@ -334,29 +368,38 @@ def local_state_dict(self) -> dict:
         """
         return self.optim.state_dict()
 
-    def state_dict(self) -> Dict[str, Any]:
+    def state_dict(self, all_ranks: bool = False) -> Dict[str, Any]:
         """Return the last known global optimizer state. The returned state is compatible with Pytorch, in that the
-        sharded properties are not exposed. It contains two entries:
+        sharded properties are not exposed.
 
-        * state - a dict holding current optimization state. Its content
-            differs between optimizer classes.
 
-        * param_groups - a dict containing all parameter groups
+        Arguments:
+            all_ranks (bool): materialize the state on all ranks. In that case, `.state_dict()` needs to be called on
+            all ranks
 
-        .. warning:
-            If the state has not been consolidated, this returns a shard's worth, not the global state.
+        Returns:
+            a dict with two entries
+                * state - a dict holding current optimization state. Its content
+                    differs between optimizer classes.
+
+                * param_groups - a dict containing all parameter groups
 
         .. warning:
-            Returning the global state is limited to the replica which was responsible for the consolidation.
-            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+            Returning the global state is limited to the replica which was responsible for the consolidation,
+            if `all_ranks` was not set to `True`. In that case, the state may also not be up to date,
+            depending on when `consolidate_state_dict` was last called.
         """
 
-        if len(self._all_states) == 0:
+        if not all_ranks and len(self._all_states) == 0:
             raise RuntimeError(
                 "Optimizer state has not been consolidated on this rank. \
                 Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
             )
```

Comment:
Would it be possible to change the runtime error to a warning ? 

The error case for us is if a user is using the sharded optimizer outside of sharded DDP. If we're not using sharded DDP, we assume the optimizer is a PyTorch optimizer. We have a simple callback that does this:
```
def on_save_checkpoint(self, trainer: Trainer, _pl_module: LightningModule) -> None:
        for optimizer in trainer.optimizers:
            if isinstance(optimizer, OSS):
                    optimizer.consolidate_state_dict()
```
After this runs, we fetch the state dict on all ranks when creating the checkpoint dict. This runs into this error on non-zero ranks. If we instead change this to become a warning and returns even an empty dictionary, then that will resolve the issue. But it depends on what the expectations are for `state_dict` and whether a warning + empty dict is a hack

Reply:
I would prefer not to, because it's not doing what the contract says it will do, it cannot return the state_dict in that case, silently failing (nobody reads warnings) feels like something which will bite the users in the future. 

I don't understand why you would fetch an empty dict in that case, since you do something differently in between the ranks anyway why not skip the state_dict() call ?

Reply:
example failure case: a user kicks a very expensive training at scale, and thinks that the checkpointing is fine. For some reason he/she is saving an empty dict from a different rank than the one which got the state, and all the checkpoints are not useable, which he/she realizes long after the training is done. I would really much prefer not to open up for a failure case like that

Reply:
Agree with @blefaudeux on this one, we've actually seen the example that Ben brought up when someone was using our pipe implementation in PL, since there is no forcing of save logic there.

As Ben highlighted, what's the use case of saving on all ranks?

Reply:
@ananthsub would that work for you ?

Reply:
ok, I just re-read 

> "The error case for us is if a user is using the sharded optimizer outside of sharded DDP. If we're not using sharded DDP, we assume the optimizer is a PyTorch optimizer.". 


The issue there is that this has nothing to do with ShardedDDP, OSS is the one sharding the optimizer state, so **with or without ShardedDDP the optimizer state is sharded**, and the vanilla Pytorch API does not work (unless we move to RPC, may be the case in the future but not really possible right now)